### DISCUSSION TITLE
Object parenting itself

### DISCUSSION THREAD

**Discussion by EduardoCardosoGit** [MarimerLLC/csla#1751] [Questions] [answered]
Hi all,

I'm having an issue where I can't to structure my object properly, the Parent property to be more precise.

 I have my AssetRE class

`[Serializable]
    public class AssetRE : BusinessBase<AssetRE>
    {
        public static readonly PropertyInfo<Guid> IdProperty = RegisterProperty<Guid>(c => c.Id);

        public Guid Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }

        public static readonly PropertyInfo<OperationsListRE> OperationsProperty = RegisterProperty<OperationsListRE>(nameof(Operations));
        public OperationsListRE Operations
        {
            get => GetProperty(OperationsProperty);
            set => LoadProperty(OperationsProperty, value);
        }

       //omit and other methods properties for brevity 
}`

OperationsListRE class

`[Serializable]
    public class OperationsListRE : BusinessBindingListBase<OperationsListRE, OperationRE>
    {
        [CreateChild]
        private void Create()
        {

        }       
            
        public void AddOperation(Guid parentId, int operationType, TimeSpan? totalTime)
        {
            using (LoadListMode)
            {
                var operation = DataPortal.Fetch<OperationRE>(parentId, operationType, totalTime);
                this.Add(operation);
            }            
        }
    }`

And OperationRE

`[Serializable]
    public class OperationRE : BusinessBase<OperationRE>
    {
        public static readonly PropertyInfo<Guid> IdProperty = RegisterProperty<Guid>(c => c.Id);

        public Guid Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }

        //omit and other methods properties for brevity       

        [Fetch]        
        private void Fetch(Guid assetId, int operationType,TimeSpan? totalTime)
        {
            Id = Guid.NewGuid();
            AssetId = assetId;
            OperationTypeId = OperationTypeEnumFunctions.OperationTypesDictionary[operationType];
            Description = ((OperationTypeEnum)operationType).ToString();
            PlannedDuration = totalTime.HasValue ? totalTime.Value.TotalMilliseconds : 0.0d;
            ParentId = assetId;            
        }

    }`

My dataflow/workflow is a bit different from what I see here in the forum. I first get data from an external source, create the rule engine objects, run the rules and then save it on the database. 

When I create the Asset I create an empty OperationsList where I'm gonna add operations later. But when I add an operation to a OperationsList the Parent on the operations is itself not the operationsList.

![image](https://user-images.githubusercontent.com/63802445/89883328-e971b080-dbbf-11ea-864e-7fdd577939d7.png)

After creating the object the parent is null, which is correct. But after adding it to the list

![image](https://user-images.githubusercontent.com/63802445/89884066-ec20d580-dbc0-11ea-8ae5-f852c6dabe3a.png)

As you can see in the image the operationRE as it self as Parent. The more operations I add more Parents I have. Shouldn't the OperationList be the parent of the Operation?

What am I doing wrong here? I'm fairly new to CSLA so I can be doing something basic thing wrong.

Thanks in advance

**Comment by hurcane** [comment]
It's showing the correct relationship. The parent of a child business object is the list (OperationsListRE). You cannot skip from the child object directly to the the AssetRE object that is two levels up the object hierarchy.

**Reply by EduardoCardosoGit** [reply]
Hi Don, 

Thanks for responding. That's what I expected, but if you check the last image you can see that the OperationRE parent is a OperationRE not a OperationListRE. 

New example with 3 operations

![image](https://user-images.githubusercontent.com/63802445/89898031-eaaed780-dbd7-11ea-9723-2ef3e843c778.png)

You can see that the father of the OperationListRE is indeed the AssetRE, but inside the OperationListRE  you have 3 OperationRE and each one of them have themselves and the "brothers" as parents, not the OperationListRE.

**Comment by rockfordlhotka** [comment]
Is the _instance_ parenting itself? Or the _type_ parenting another instance of the same type?

**Reply by EduardoCardosoGit** [reply]
Hi Rockford, thank you so much for taking time to help me, I know you are a very busy person.

The instance is parenting itself.

If I have:

OperationsListRE
   &nbsp;    -Operation 1
    &nbsp;   -Operation 2
    &nbsp;   -Operation 3

The Parent of those 3 operations is populated with the same 3 operations

OperationsListRE
   &nbsp;    Operation 1
       &nbsp;      - Parent
                &nbsp; &nbsp; &nbsp;   - Operation 1
                &nbsp; &nbsp; &nbsp;    -Operation 2
               &nbsp; &nbsp; &nbsp;     -Operation 3 
     &nbsp;  Operation 2
             &nbsp;   - Parent
                  &nbsp; &nbsp; &nbsp;    -Operation 1
                 &nbsp; &nbsp; &nbsp;     -Operation 2
                 &nbsp; &nbsp; &nbsp;     -Operation 3 
     &nbsp;  Operation 3
              -Parent
                 &nbsp; &nbsp; &nbsp;     -Operation 1
                 &nbsp; &nbsp; &nbsp;     -Operation 2
                 &nbsp; &nbsp; &nbsp;     -Operation 3 

Thanks again.

**Reply by rockfordlhotka** [reply]
So `OperationsListRE` is a `BusinessListBase`, and `Operation` is also a `BusinessListBase`? Do you have some leaf node type that has properties?

**Reply by EduardoCardosoGit** [reply]
No, OperationsListRE is a BusinessListBase, and Operation is a BusinessBase. They all have properties, I omitted them just for the sake of simplicity

**Reply by rockfordlhotka** [reply]
How does `Operation` contain a list of child `Operation` objects?

Something like this?

`OperationsListRE` -list-of-> `Operation` -contains-> `OperationListRE`

**Comment by EduardoCardosoGit** [comment]
It starts with Asset

```c#
 [Serializable]
    public class AssetRE : BusinessBase [AssetRE] (the edit didnt like the <>here)
    {
        public static readonly PropertyInfo<Guid> IdProperty = RegisterProperty<Guid>(c => c.Id);

        public Guid Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }

        //omitting properties

        public static readonly PropertyInfo<OperationsListRE> OperationsProperty = RegisterProperty<OperationsListRE>(nameof(Operations));
        public OperationsListRE Operations
        {
            get => GetProperty(OperationsProperty);
            set => LoadProperty(OperationsProperty, value);
        }`

Then OperationsListRE which is a property of AssetRE

```c#
[Serializable]
    public class OperationsListRE : BusinessBindingListBase<OperationsListRE, OperationRE>
    {}
```

Then OperationRE

```c#
 [Serializable]
    public class OperationRE : BusinessBase[OperationRE] (the edit didnt like the <>here)
    {
        public static readonly PropertyInfo<Guid> IdProperty = RegisterProperty<Guid>(c => c.Id);

        public Guid Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }
     //omitting properties
}
```
So, it's:

-AssetRE
&nbsp; -OperationListRE
&nbsp; &nbsp; -OperationRE 1
&nbsp; &nbsp; -OperationRE 2
&nbsp; &nbsp; -etc

Asset has list of operations.

I can see the Parent of OperationsListRE being AssetRE as it should but when I check the Parent property of any of the OperationRE it I see the object itself + it's "brothers". Like on this image

![image](https://user-images.githubusercontent.com/63802445/89921284-6703e380-dbf5-11ea-9b62-8cd50a524a26.png)

I thought that the OperationRE parent would be the OperationsListRE, shouldn't it be?

**Reply by rockfordlhotka** [reply]
CSLA manages the `Parent` reference automatically, and it is set when you add an object to a collection or as a child of another object. An object can only have one parent, and there's no defined behavior for setting the parent more than one time. I _am guessing_ that which ever object you added the child to last will be the "parent".

**Reply by rockfordlhotka** [reply]
Logically what you have is an object graph, and in that graph you have an object that needs to reference an arbitrary child (which might be itself) right?

That reference should be implemented differently, because you aren't establishing a parent-child relationship, just a reference.

```c#
using Csla;
using System;
using System.Linq;

namespace ConsoleApp4
{
  [Serializable]
  public class OperationList : BusinessListBase<OperationList, Operation>
  {
    [Fetch]
    private void Fetch()
    {
      using (LoadListMode)
      {
        Add(DataPortal.CreateChild<Operation>(123));
        Add(DataPortal.CreateChild<Operation>(456));
        Add(DataPortal.CreateChild<Operation>(789));
      }
    }
  }

  [Serializable]
  public class Operation : BusinessBase<Operation>
  {
    public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(nameof(Id));
    public int Id
    {
      get => GetProperty(IdProperty);
      set => SetProperty(IdProperty, value);
    }

    public static readonly PropertyInfo<int> SubOperationIdProperty = RegisterProperty<int>(nameof(SubOperationId));
    public int SubOperationId
    {
      get => GetProperty(SubOperationIdProperty);
      set => SetProperty(SubOperationIdProperty, value);
    }

    public Operation SubOperation
    {
      get
      {
        var p = (OperationList)Parent;
        return p.Where(r => r.Id == SubOperationId).FirstOrDefault();
      }
    }

    public static readonly PropertyInfo<OperationList> SubOperationsProperty = RegisterProperty<OperationList>(nameof(SubOperations));
    public OperationList SubOperations
    {
      get => GetProperty(SubOperationsProperty);
      private set => LoadProperty(SubOperationsProperty, value);
    }

    [FetchChild]
    private void Fetch(int id)
    {
      using (BypassPropertyChecks)
      {
        Id = id;
      }
    }
  }
}
```

**Reply by EduardoCardosoGit** [reply]
That was my expectation, but that's not what happens. When I add a child (OperationRE) to the father (OperationsListRE) the parent 

public void AddOperation(Guid parentId, int operationType, TimeSpan? totalTime)
{
            using (LoadListMode)
            {
                var operation = DataPortal.FetchChild<OperationRE>(parentId, operationType, totalTime);
                this.Add(operation);
            }
}

The parent of the OperationRE is set to itself as I showed on the images above. And as more a I add the more Parents I have.

Can I send you a small POC showing the issue?

**Reply by EduardoCardosoGit** [reply]
Sorry, responded without checking your 2nd message.

Checking it now

**Reply by EduardoCardosoGit** [reply]
I may have explain myself badly, sorry for that. Let me use your example, If create an operation with 3 sub operations just like your code did, what I thought it would happen was:

Operation.Parent=null which is correct, it's the root.

Operation.SubOperations.Parent = Operation

and finally

Operation.SubOperations[0].Parent = Operation.SubOperations
Operation.SubOperations[1].Parent = Operation.SubOperations
Operation.SubOperations[2].Parent = Operation.SubOperations

all the 3 childs would have the same Parent.

but what happens is 

Operation.SubOperations[0].Parent = List<Operation.SubOperations> the same as in Operation.SubOperations
Operation.SubOperations[1].Parent = List<Operation.SubOperations> the same as in Operation.SubOperations
Operation.SubOperations[2].Parent = List<Operation.SubOperations> the same as in Operation.SubOperations

Is this expected behaviour? From what you and @hurcane  are saying it is and I'm missing something...

Sending the console app file for reference
[Program.zip](https://github.com/MarimerLLC/csla/files/5058807/Program.zip)

**Comment by hurcane** [comment]
It all looks correct to me. I think some of the Parent references in the watch window are being interpreted incorrectly.

this is the list.
this.Parent is the Asset
then you have expanded this[0], which is Operation1
the Parent that you highlighted is Operation1.Parent, which points back to the list (i.e. this)

Any graph with the child referencing the parent can be expanded in a circular fashion. If you override ToString on OperationRE and return the identity value for the operation, you will see that you are just showing a circular reference within the Watch window.

Another test: 
ReferenceEquals(this, this[0].Parent) - will be true
ReferenceEquals(this, this[0].Parent[0].Parent) - will be true

**Comment by hurcane** [comment, marked as answer]
I ran the code and here are the results from some of the parent references that I added to the watch window.


  | Name | Value | Type
-- | -- | -- | --
▶ | operation.SubOperations[0].Parent | Count = 3 | Csla.Core.IParent {RockTest.OperationList}
▶ | (OperationList)(operation.SubOperations[0].Parent) | Count = 3 | RockTest.OperationList
▶ | (OperationList)((OperationList)(operation.SubOperations[0].Parent))[2].Parent | Count = 3 | RockTest.OperationList
  | ReferenceEquals((OperationList)(operation.SubOperations[0].Parent),(OperationList)((OperationList)(operation.SubOperations[0].Parent))[2].Parent) | true | bool

I noted that the "raw" Parent result is an interface, which you can see in the first row.

The second row is the same as the first, but cast back to the business class.

The third row is a reference to operation.SubOperations[0].Parent[2].Parent. You cannot actually use that in the debugger because each Parent reference needs to cast back to the base class to access the children, but I'm effectively going to child 0, back to the parent, down to child 2, and back to the parent. 

The final row proves that they are indeed the same instance in memory.

The debugger has trouble working with these kinds of circular references. For instance, I can right-click and "Add watch" on the top Parent, and the type returned is IParent. I can also add watch for the Parent of the sub operation, but the debugger generates its own generic debug view collection from the contents and things start to get confused.

I don't thinks this is a fault of Csla. I think this is a debugger visualization limitation.

**Reply by EduardoCardosoGit** [reply]
So odd...I'm fairly new to CSLA so I thought I was doing something wrong.

As long as the rules work fine, accessing the parent, firing up child changes events,that kind of stuff I'm good with it. I'm going to do some tests around that.

Many thanks for your help, much appreciated. Of course @rockfordlhotka also, thanks for taking the time.
